html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/OAuth/Instagram.pm

```diff
@@ -0,0 +1,178 @@
+package captiveportal::PacketFence::DynamicRouting::Module::Authentication::OAuth::Instagram;
+
+=head1 NAME
+
+captiveportal::DynamicRouting::Module::Authentication::OAuth::Instagram
+
+=head1 DESCRIPTION
+
+Instagram OAuth module
+
+=cut
+
+use WWW::Curl::Easy;
+use Moose;
+use pf::log;
+extends 'captiveportal::DynamicRouting::Module::Authentication::OAuth';
+
+has '+source' => (isa => 'pf::Authentication::Source::InstagramSource');
+
+=head2 get_token
+
+Get the OAuth2 token
+
+=cut
+
+sub get_token {
+    my ($self) = @_;
+
+    my $curl = WWW::Curl::Easy->new;
+    my $code = $self->app->request->parameters->{code};
+    my $info = $self->get_client;
+
+    my $source = $self->source;
+
+    my $response_body = '';
+    my $formdata = WWW::Curl::Form->new;
+    
+    $formdata->formadd("client_id", "$info->{NOP_id}");
+    $formdata->formadd("client_secret", "$info->{NOP_secret}");
+    $formdata->formadd("redirect_uri", "$info->{NOPW_redirect}");
+    $formdata->formadd("grant_type", "authorization_code");
+    $formdata->formadd("code", "$code");
+
+    $curl->setopt(CURLOPT_HTTPPOST, $formdata);
+    $curl->setopt(CURLOPT_WRITEDATA, \$response_body);
+    $curl->setopt(CURLOPT_DNS_USE_GLOBAL_CACHE, 0);
+    $curl->setopt(CURLOPT_NOSIGNAL, 1);
+    $curl->setopt(CURLOPT_HTTPHEADER(), ['Content-Type: multipart/form-data']);
+    $curl->setopt(CURLOPT_URL, $info->{NOP_access_token_url});
+
+    my $curl_return_code = $curl->perform;
+
+    my $response_code = $curl->getinfo(CURLINFO_HTTP_CODE);
+
+    use JSON;
+    my $json = JSON->new;
+    my $jsond = $json->decode($response_body);
+
+    my $token = $jsond->{access_token};
+
+    if ($@) {
+        get_logger->warn("OAuth2: failed to receive the token from the provider: $@");
+        pf::auth_log::change_record_status($self->source->id, $self->current_mac, $pf::auth_log::FAILED);
+        $self->app->flash->{error} = "OAuth2 Error: Failed to get the token";
+        $self->landing();
+        return;
+    }
+    return $token;
+
+}
+
+=head2 handle_callback
+
+Handle the callback from the OAuth2 provider and fetch the protected resource
+
+=cut
+
+sub handle_callback {
+    my ($self) = @_;
+
+    my $token = $self->get_token();
+    return unless($token);
+    my $info = $self->get_client;
+
+    # request a JSON response
+
+    my $curl = WWW::Curl::Easy->new;
```

Comment:
Create yourself a helper method to create your curl object with the options that are common to both calls